### PR TITLE
 Add support for HTTP(S) OTLP endpoint 

### PR DESCRIPTION
**Pull request by sharpSteff**
Hi Aspire-Team,

Great work with the Aspire-Dashboard. 
I'm currently looking for a replacement for current OTLP-Backend (https://github.com/SigNoz/signoz), and Aspire Dashboard seems to be a lightweight sulution.

However we are relying in the HTTP OTLP Receiver (#3688). Therefore i tried my best to add HTTP OTLP endpoint support and it works quite will in our usecase.

With this changes Aspire Dashboard can consume OTLP Grpc and http(s) simultaneously. Due to the http/protobuf I could reuse the existing grpc-otlps models and services

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/4197)

### CODE REVIEW

File: src/Aspire.Dashboard/Otlp/WebApplicationExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    public static WebApplication ConfigureHttpOtlp(this WebApplication app, Uri httpEndpoint)
+    {
+        app.MapPost("/v1/metrics",
+            async ([FromServices] OtlpMetricsService service, HttpContext httpContext) =>
+            {
+                var body = await ReadFullyAsync(httpContext).ConfigureAwait(false);
+                var metrics = ExportMetricsServiceRequest.Parser.ParseFrom(body);
+                await service.Export(metrics, null!).ConfigureAwait(false);
+            }
+        ).AllowAnonymous()
```
Comment (davidfowl) [inline src/Aspire.Dashboard/Otlp/WebApplicationExtensions.cs:23]: Shouldn't this use the otlp auth policy?

https://github.com/dotnet/aspire/blob/03c6a503930f0a37a089a05fa82b49e636b261dd/src/Aspire.Dashboard/Otlp/Grpc/OtlpMetricsService.cs#L14

File: src/Aspire.Dashboard/Otlp/WebApplicationExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+        return app;
+    }
+
+    private static async Task<byte[]> ReadFullyAsync(HttpContext context)
+    {
+        using var ms = new MemoryStream();
+        await context.Request.Body.CopyToAsync(ms).ConfigureAwait(false);
+        return ms.ToArray();
+    }
```
Comment (davidfowl) [inline src/Aspire.Dashboard/Otlp/WebApplicationExtensions.cs:54]: This should be using pipelines to buffer without allocating.

**Comment by davidfowl** [comment]
The formatting changes make it really hard to review. I don't know if you can undo those but with a big change like this it really makes the diff bigger than it needs to be.

File: src/Aspire.Dashboard/Otlp/WebApplicationExtensions.cs (hunk 1 above)
Comment (sharpSteff) [reply src/Aspire.Dashboard/Otlp/WebApplicationExtensions.cs]: Good catch!
changed to 
`.RequireAuthorization(OtlpAuthorization.PolicyName)`

**Comment by sharpSteff** [comment]
> The formatting changes make it really hard to review. I don't know if you can undo those but with a big change like this it really makes the diff bigger than it needs to be.

I reverted the format changes. Sorry for that

File: src/Aspire.Dashboard/Otlp/WebApplicationExtensions.cs (hunk 2 above)
Comment (sharpSteff) [reply src/Aspire.Dashboard/Otlp/WebApplicationExtensions.cs:55]: How do I achieve this?
`Google.Protobuf.MessageParser<T>` wants either `ReadOnlySequence<byte> `, `ReadOnlySpan<byte>` or a `Stream`. but the HttpContext.Request.Body-Stream can not be passed directly

Comment (sharpSteff) [reply src/Aspire.Dashboard/Otlp/WebApplicationExtensions.cs:55]: like this?
```
var reader = httpContext.Request.BodyReader;
var readResult = await reader.ReadAsync().ConfigureAwait(false);
var readResultBuffer = readResult.Buffer;
var metrics = ExportMetricsServiceRequest.Parser.ParseFrom(readResultBuffer);
```

Comment (sharpSteff) [reply src/Aspire.Dashboard/Otlp/WebApplicationExtensions.cs]: @davidfowl I got rid of the byte[] and stream allocation

**Comment by eerhardt** [comment]
Are there tests that can be written for this new functionality?

File: src/Aspire.Dashboard/Otlp/WebApplicationExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        var readResult = await reader.ReadAsync().ConfigureAwait(false);
+        var readResultBuffer = readResult.Buffer;
+        if (readResultBuffer.IsEmpty && readResult.IsCompleted)
+        {
+            return;
+        }
+
+        await exporter(readResultBuffer).ConfigureAwait(false);
+        reader.AdvanceTo(readResultBuffer.Start, readResultBuffer.End);
+        await reader.CompleteAsync().ConfigureAwait(false);
```
Comment (davidfowl) [inline src/Aspire.Dashboard/Otlp/WebApplicationExtensions.cs:57]: I would suggest something like this:

https://github.com/DamianEdwards/MinimalApis.Extensions/blob/f1a6ab59edadd7625d12770adbede0569ed90244/src/MinimalApis.Extensions/Binding/BodyOfT.cs#L150-L171

File: src/Aspire.Dashboard/Configuration/DashboardOptions.cs [hunk 4]
```diff
@@ -73,10 +74,18 @@ public sealed class OtlpOptions
     public OtlpAuthMode? AuthMode { get; set; }
     public string? EndpointUrl { get; set; }
```
Comment (drewnoakes) [inline src/Aspire.Dashboard/Configuration/DashboardOptions.cs:75]: Can we rename this to `GrpcEndpointUrl` for consistency?

**Review by drewnoakes — COMMENTED** [review]
Looks like a good addition, thanks. Left a few comments.

It'd be great to add a playground application that exercised these options, so there was an F5-able test bed, for manual testing.

There are integration tests for the current gRPC-based OTLP interactions. Having similar tests for the HTTP interactions would be valuable to ensure this scenario doesn't break.

File: src/Aspire.Dashboard/Otlp/WebApplicationExtensions.cs (hunk 3 above)
Comment (sharpSteff) [reply src/Aspire.Dashboard/Otlp/WebApplicationExtensions.cs]: done

File: src/Aspire.Dashboard/Configuration/DashboardOptions.cs (hunk 4 above)
Comment (sharpSteff) [reply src/Aspire.Dashboard/Configuration/DashboardOptions.cs]: done

**Comment by sharpSteff** [comment]
@eerhardt @drewnoakes 
> Are there tests that can be written for this new functionality?
I added for all endpoints IntegrationTests

File: src/Aspire.Dashboard/DashboardWebApplication.cs [hunk 5]
```diff
@@ -43,6 +44,7 @@ public sealed class DashboardWebApplication : IAsyncDisposable
     private readonly IReadOnlyList<string> _validationFailures;
     private Func<EndpointInfo>? _frontendEndPointAccessor;
     private Func<EndpointInfo>? _otlpServiceEndPointAccessor;
```
Comment (JamesNK) [inline src/Aspire.Dashboard/DashboardWebApplication.cs:46]: Can you change:

* _otlpServiceEndPointAccessor -> _otlpServiceGrpcEndPointAccessor
* OtlpServiceEndPointAccessor -> OtlpServiceGrpcEndPointAccessor

If there are two OTLP endpoint types then the variables should always say whether it's the grpc or http.

File: src/Aspire.Dashboard/DashboardWebApplication.cs [hunk 6]
```diff
@@ -339,7 +359,12 @@ private void ConfigureKestrelEndpoints(WebApplicationBuilder builder, DashboardO
         {
             // Translate high-level config settings such as DOTNET_DASHBOARD_OTLP_ENDPOINT_URL and ASPNETCORE_URLS
             // to Kestrel's schema for loading endpoints from configuration.
-            AddEndpointConfiguration(initialValues, "Otlp", otlpUri.OriginalString, HttpProtocols.Http2, requiredClientCertificate: dashboardOptions.Otlp.AuthMode == OtlpAuthMode.ClientCertificate);
+            AddEndpointConfiguration(initialValues, "Otlp", otlpGrpcUri.OriginalString, HttpProtocols.Http2, requiredClientCertificate: dashboardOptions.Otlp.AuthMode == OtlpAuthMode.ClientCertificate);
```
Comment (JamesNK) [inline src/Aspire.Dashboard/DashboardWebApplication.cs:362]:
```suggestion
            AddEndpointConfiguration(initialValues, "OtlpGrpc", otlpGrpcUri.OriginalString, HttpProtocols.Http2, requiredClientCertificate: dashboardOptions.Otlp.AuthMode == OtlpAuthMode.ClientCertificate);
```

There might be some tests that depend on setting the name. Technically this is a breaking change from external config perspective, but we don't document anywhere that endpoints have names so I don't consider it part of the API.

File: src/Aspire.Dashboard/DashboardWebApplication.cs [hunk 7]
```diff
@@ -357,7 +382,7 @@ private void ConfigureKestrelEndpoints(WebApplicationBuilder builder, DashboardO
         }
         else
         {
-            AddEndpointConfiguration(initialValues, "Otlp", otlpUri.OriginalString, HttpProtocols.Http1AndHttp2, requiredClientCertificate: dashboardOptions.Otlp.AuthMode == OtlpAuthMode.ClientCertificate);
+            AddEndpointConfiguration(initialValues, "Otlp", otlpGrpcUri.OriginalString, HttpProtocols.Http1AndHttp2, requiredClientCertificate: dashboardOptions.Otlp.AuthMode == OtlpAuthMode.ClientCertificate);
```
Comment (JamesNK) [inline src/Aspire.Dashboard/DashboardWebApplication.cs:385]:
```suggestion
            AddEndpointConfiguration(initialValues, "OtlpGrpc", otlpGrpcUri.OriginalString, HttpProtocols.Http1AndHttp2, requiredClientCertificate: dashboardOptions.Otlp.AuthMode == OtlpAuthMode.ClientCertificate);
```

File: src/Aspire.Dashboard/Otlp/WebApplicationExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+
+public static class WebApplicationExtensions
+{
+    // https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/large-object-heap
+    private const int MaxSizeLessThanLOH = 84999;
+
+    public static WebApplication ConfigureHttpOtlp(this WebApplication app, Uri httpEndpoint)
+    {
+        app.MapPost("/v1/metrics",
+                async ([FromServices] OtlpMetricsService service, HttpContext httpContext) =>
```
Comment (JamesNK) [inline src/Aspire.Dashboard/Otlp/WebApplicationExtensions.cs:22]: Is this injecting the OTLP gRPC service into the OTLP HTTP API? I haven't looked closely into what this means, but on the surface I'm not sure I like that coupling.

File: src/Aspire.Dashboard/Otlp/WebApplicationExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+                {
+                    return await ExportOtlpData(httpContext,
+                            sequence => service.Export(ExportMetricsServiceRequest.Parser.ParseFrom(sequence), null!))
+                        .ConfigureAwait(false);
```
Comment (JamesNK) [inline src/Aspire.Dashboard/Otlp/WebApplicationExtensions.cs:26]: Is this API accepting a protobuf request message and then returning a JSON response message? I couldn't see any Protobuf serialization code so I assume that the MapPost func is returning a value and ASP.NET Core is doing the serialization, which is always JSON.

I need to read the specification for OTLP+HTTP. I know it supports both Protobuf and JSON, but both in one request seems very weird. I think to support both the API method would have to look at the request content-type and do some content negotiation. I've yet to do that sort of thing inside minimal API methods so I don't know how hard/easy it is.

Also, I'm surprised the Protobuf response message just works when serialized to JSON if the .NET serializer is just accepting a plain old C# class and serializing it. Protobuf has rules about how it maps Protobuf messages to JSON (and unfortunately OTLP+HTTP/JSON breaks some of them). Maybe it works because the export response message is so simple.

File: src/Aspire.Dashboard/Otlp/WebApplicationExtensions.cs [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+                }
+            ).RequireAuthorization(OtlpAuthorization.PolicyName)
+            .RequireHost($"*:{httpEndpoint.Port}");
```
Comment (JamesNK) [inline src/Aspire.Dashboard/Otlp/WebApplicationExtensions.cs:29]: I used connection middleware to ensure OTLP + gRPC request came in on the right endpoint. I'd like OTLP + HTTP to be consistent.

There are probably some relatively modest changes/improvements to reuse the same code for both.

**Review by JamesNK — CHANGES_REQUESTED** [review]
Some small nits but also big issues. I need to read the OTLP HTTP spec to understand what exactly we need to support.

I'm on leave at the moment, but when I get back (beginning of June) are you ok with me making changes to your PR? I think communicating everything via comments will be too much overhead.

File: src/Aspire.Dashboard/Otlp/WebApplicationExtensions.cs (hunk 9 above)
Comment (JamesNK) [reply src/Aspire.Dashboard/Otlp/WebApplicationExtensions.cs:26]: I think it will be much easier to initially only support OTLP+HTTP with Protobuf. Supporting JSON will require a lot of custom JSON serialization.

**Comment by sharpSteff** [comment]
@JamesNK It's a big pleasure to receive a review of Newtonsoft.Json's author. Thanks for all your work!

File: src/Aspire.Dashboard/Otlp/WebApplicationExtensions.cs (hunk 9 above)
Comment (sharpSteff) [reply src/Aspire.Dashboard/Otlp/WebApplicationExtensions.cs:26]: The OTLP .NET SDK  supports only **HTTP/Protobuf** and **grpc** (https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Exporter.OpenTelemetryProtocol/README.md). I suggest to add HTTP/JSON support when OLTP .NET does also support it.

Comment (sharpSteff) [reply src/Aspire.Dashboard/Otlp/WebApplicationExtensions.cs:26]: Regarding the return value, you a probably right. I have to check the spec here.
The OTLP .NET HTTP Exporter pretty much just checks the status code of the response and stores the headers in bad case:

https://github.com/open-telemetry/opentelemetry-dotnet/blob/043d8a31c768d05544c06629e072492238a0069b/src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ExportClient/BaseOtlpHttpExportClient.cs#L43-L57

File: src/Aspire.Dashboard/Otlp/WebApplicationExtensions.cs (hunk 8 above)
Comment (sharpSteff) [reply src/Aspire.Dashboard/Otlp/WebApplicationExtensions.cs:22]: I'm with you. I would like to have a protocol indipendent service to export otlp data

File: src/Aspire.Dashboard/Otlp/WebApplicationExtensions.cs (hunk 10 above)
Comment (sharpSteff) [reply src/Aspire.Dashboard/Otlp/WebApplicationExtensions.cs:28]: @JamesNK any idea how to make this work elegantly ?

File: src/Aspire.Dashboard/Otlp/WebApplicationExtensions.cs [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+using OpenTelemetry.Proto.Collector.Trace.V1;
+
+namespace Aspire.Dashboard.Otlp;
+
+public static class WebApplicationExtensions
+{
+    // https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/large-object-heap
+    private const int MaxSizeLessThanLOH = 84999;
+
+    public static WebApplication ConfigureHttpOtlp(this WebApplication app, Uri httpEndpoint)
```
Comment (wertzui) [inline src/Aspire.Dashboard/Otlp/WebApplicationExtensions.cs:18]: I think, you have to configure CORS too, so browsers can send data to this endpoint.

Comment (davidfowl) [reply src/Aspire.Dashboard/Otlp/WebApplicationExtensions.cs:18]: Lets avoid the cors configuration until we understand what it means to do this securely.

**Comment by JamesNK** [comment]
I made a bunch of changes.

* Locked the endpoint down to Protobuf payloads
* Make OTLP/gRPC endpoint optional. Either OTLP/gRPC or OTLP/HTTP is required.
* More tests

File: src/Aspire.Dashboard/Otlp/Http/OtlpHttpEndpointsBuilder.cs [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+
+    public static void ConfigureHttpOtlp(this IEndpointRouteBuilder endpoints, Uri httpEndpoint)
+    {
+        var logsService = endpoints.ServiceProvider.GetRequiredService<OtlpLogsService>();
+        var traceService = endpoints.ServiceProvider.GetRequiredService<OtlpTraceService>();
+        var metricsService = endpoints.ServiceProvider.GetRequiredService<OtlpMetricsService>();
+
+        endpoints.MapPost("/v1/logs", LogsEndpoint).AddOtlpHttpMetadata();
+        endpoints.MapPost("/v1/traces", TracesEndpoint).AddOtlpHttpMetadata();
+        endpoints.MapPost("/v1/metrics", MetricsEndpoint).AddOtlpHttpMetadata();
```
Comment (davidfowl) [inline src/Aspire.Dashboard/Otlp/Http/OtlpHttpEndpointsBuilder.cs:31]:
```suggestion
        var group =  endpoints.MapGroup("/v1").AddOtlpHttpMetadata();
        
        group.MapPost("logs", LogsEndpoint);
        group.MapPost("traces", TracesEndpoint);
        group.MapPost("metrics", MetricsEndpoint);
```

File: src/Aspire.Dashboard/Otlp/Http/OtlpHttpEndpointsBuilder.cs [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+                    break;
+                case KnownContentType.Json:
+                default:
+                    context.Response.StatusCode = StatusCodes.Status415UnsupportedMediaType;
+                    break;
+            }
+        }
+    }
+
+    private static async Task WriteOtlpResponse<T>(HttpContext context, T response) where T : IMessage
```
Comment (davidfowl) [inline src/Aspire.Dashboard/Otlp/Http/OtlpHttpEndpointsBuilder.cs:83]: These are small responses right?

File: src/Aspire.Dashboard/Otlp/Http/OtlpHttpEndpointsBuilder.cs [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+    // https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/large-object-heap
+    private const int MaxSizeLessThanLOH = 84999;
+    public const string ProtobufContentType = "application/x-protobuf";
+    public const string JsonContentType = "application/json";
+
+    public static void ConfigureHttpOtlp(this IEndpointRouteBuilder endpoints, Uri httpEndpoint)
+    {
+        var logsService = endpoints.ServiceProvider.GetRequiredService<OtlpLogsService>();
+        var traceService = endpoints.ServiceProvider.GetRequiredService<OtlpTraceService>();
+        var metricsService = endpoints.ServiceProvider.GetRequiredService<OtlpMetricsService>();
```
Comment (davidfowl) [inline src/Aspire.Dashboard/Otlp/Http/OtlpHttpEndpointsBuilder.cs:27]: These should be injected into the endpoints.

File: src/Aspire.Dashboard/Otlp/Http/OtlpHttpEndpointsBuilder.cs (hunk 13 above)
Comment (JamesNK) [reply src/Aspire.Dashboard/Otlp/Http/OtlpHttpEndpointsBuilder.cs:83]: Yes. Usually just a number. Maybe also a string if there was an error.

**Comment by JamesNK** [comment]
I think everything is done and this is mergable.

A follow-up piece of work is changing the app host templates to include `DOTNET_DASHBOARD_OTLP_HTTP_ENDPOINT_URL` env var. Will do that in its own PR as I think it merits its own discussion without distraction.

File: src/Aspire.Dashboard/Otlp/Http/OtlpHttpEndpointsBuilder.cs [hunk 15]
```diff
@@ -0,0 +1,200 @@
...
+using OpenTelemetry.Proto.Collector.Trace.V1;
+
+namespace Aspire.Dashboard.Otlp.Http;
+
+public static class OtlpHttpEndpointsBuilder
+{
+    public const string ProtobufContentType = "application/x-protobuf";
+    public const string JsonContentType = "application/json";
+
+    public static void ConfigureHttpOtlp(this IEndpointRouteBuilder endpoints)
```
Comment (davidfowl) [inline src/Aspire.Dashboard/Otlp/Http/OtlpHttpEndpointsBuilder.cs:22]: nit: 
```suggestion
    public static void MapOtlpEndpoints(this IEndpointRouteBuilder endpoints)
```

Map* is the verb we use for adding routes. Configure is for options.

Comment (JamesNK) [reply src/Aspire.Dashboard/Otlp/Http/OtlpHttpEndpointsBuilder.cs:22]: I think I came up with the `Map*` verb back in .NET Core 3 😋

File: src/Aspire.Dashboard/Otlp/Http/OtlpHttpEndpointsBuilder.cs [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+            {
+                return await next(context).ConfigureAwait(false);
+            }
+            catch (BadHttpRequestException ex)
+            {
+                _logger.LogError(ex, "Bad HTTP request when receiving OTLP data.");
+                return Results.BadRequest(ex.Message);
+            }
+        }
+    }
```
Comment (davidfowl) [inline src/Aspire.Dashboard/Otlp/Http/OtlpHttpEndpointsBuilder.cs:134]:
```suggestion
    private sealed class ErrorHandlerEndpointFilter(ILogger<ErrorHandlerEndpointFilter> logger) : IEndpointFilter
    {
        public async ValueTask<object?> InvokeAsync(EndpointFilterInvocationContext context, EndpointFilterDelegate next)
        {
            try
            {
                return await next(context).ConfigureAwait(false);
            }
            catch (BadHttpRequestException ex)
            {
                logger.LogError(ex, "Bad HTTP request when receiving OTLP data.");
                return Results.BadRequest(ex.Message);
            }
        }
    }
```

File: src/Aspire.Dashboard/Otlp/Grpc/OtlpGrpcLogsService.cs [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+using Microsoft.AspNetCore.Mvc;
+using OpenTelemetry.Proto.Collector.Logs.V1;
+
+namespace Aspire.Dashboard.Otlp.Grpc;
+
+[Authorize(Policy = OtlpAuthorization.PolicyName)]
+[SkipStatusCodePages]
+public class OtlpGrpcLogsService : LogsService.LogsServiceBase
+{
+    private readonly ILogger<OtlpGrpcLogsService> _logger;
```
Comment (drewnoakes) [inline src/Aspire.Dashboard/Otlp/Grpc/OtlpGrpcLogsService.cs:16]: `_logger` is unused.

This class is a good candidate for a primary constructor.

Ditto: `OtlpGrpcMetricsService` and `OtlpGrpcTraceService`.

File: src/Aspire.Dashboard/Otlp/Http/OtlpHttpEndpointsBuilder.cs [hunk 18]
```diff
@@ -0,0 +1,205 @@
...
+        None,
+        Protobuf,
+        Json
+    }
+
+    private static KnownContentType GetKnownContentType(string? contentType, out StringSegment charSet)
+    {
+        if (contentType != null && MediaTypeHeaderValue.TryParse(contentType, out var mt))
+        {
+            if (string.Equals(mt.MediaType, JsonContentType, StringComparison.OrdinalIgnoreCase))
```
Comment (drewnoakes) [inline src/Aspire.Dashboard/Otlp/Http/OtlpHttpEndpointsBuilder.cs:66]: Could introduce `StringComparisons.MediaType` for this and the comparison below.

File: tests/Aspire.Dashboard.Tests/Integration/IntegrationTestHelpers.cs [hunk 19]
```diff
@@ -78,6 +79,30 @@ public static class IntegrationTestHelpers
         return dashboardWebApplication;
     }
 
+    public static HttpClient CreateHttpClient(
+        string address,
+        Action<X509Certificate2?>? validationCallback = null,
+        X509CertificateCollection? clientCertificates = null)
+    {
+        var handler = new SocketsHttpHandler
```
Comment (drewnoakes) [inline tests/Aspire.Dashboard.Tests/Integration/IntegrationTestHelpers.cs:87]: For the resource service client, we allow configuration of SSL options via `configuration.Bind("Dashboard:ResourceServiceClient:Ssl", httpHandler.SslOptions);`. Do we want something similar here?

File: src/Aspire.Dashboard/Configuration/DashboardOptions.cs [hunk 20]
```diff
@@ -89,18 +96,22 @@ public byte[] GetPrimaryApiKeyBytes()
 
     internal bool TryParseOptions([NotNullWhen(false)] out string? errorMessage)
     {
-        if (string.IsNullOrEmpty(EndpointUrl))
+        if (string.IsNullOrEmpty(GrpcEndpointUrl) && string.IsNullOrEmpty(HttpEndpointUrl))
         {
-            errorMessage = $"OTLP endpoint URL is not configured. Specify a {DashboardConfigNames.DashboardOtlpUrlName.EnvVarName} value.";
+            errorMessage = $"Both OTLP gRPC endpoint URL and OTLP HTTP endpoint URL are not configured. Specify either a {DashboardConfigNames.DashboardOtlpGrpcUrlName.EnvVarName} or {DashboardConfigNames.DashboardOtlpHttpUrlName.EnvVarName} value.";
```
Comment (drewnoakes) [inline src/Aspire.Dashboard/Configuration/DashboardOptions.cs:101]: I found this a bit hard to parse. What about this?

```suggestion
            errorMessage = $"Neither gRPC or HTTP endpoint URL are configured for OTLP. Specify either a {DashboardConfigNames.DashboardOtlpGrpcUrlName.EnvVarName} or {DashboardConfigNames.DashboardOtlpHttpUrlName.EnvVarName} value.";
```

**Review by drewnoakes — APPROVED** [review]

File: tests/Aspire.Dashboard.Tests/Integration/IntegrationTestHelpers.cs (hunk 19 above)
Comment (JamesNK) [reply tests/Aspire.Dashboard.Tests/Integration/IntegrationTestHelpers.cs:87]: This is just test code. In the real world, the client in this situation is the OTEL SDK.

File: src/Aspire.Dashboard/Otlp/Http/OtlpHttpEndpointsBuilder.cs (hunk 18 above)
Comment (JamesNK) [reply src/Aspire.Dashboard/Otlp/Http/OtlpHttpEndpointsBuilder.cs:66]: I think this is a one off usage.

**Review by JamesNK — APPROVED** [review]

File: src/Aspire.Dashboard/Otlp/Http/OtlpHttpEndpointsBuilder.cs [hunk 21]
```diff
@@ -0,0 +1,205 @@
...
+using OpenTelemetry.Proto.Collector.Trace.V1;
+
+namespace Aspire.Dashboard.Otlp.Http;
+
+public static class OtlpHttpEndpointsBuilder
+{
+    public const string ProtobufContentType = "application/x-protobuf";
+    public const string JsonContentType = "application/json";
+
+    public static void MapHttpOtlpApi(this IEndpointRouteBuilder endpoints)
```
Comment (captainsafia) [inline src/Aspire.Dashboard/Otlp/Http/OtlpHttpEndpointsBuilder.cs:23]: It's typically recommended that `MapX` implementations like this one return the `IEndpoitRouteBuilder` instance passed in in order to support chaining invocations onto the call (e.g. adding metadata via `WithMetadta`). Is there a particular reason to prohibit chaining on this API?

Comment (davidfowl) [reply src/Aspire.Dashboard/Otlp/Http/OtlpHttpEndpointsBuilder.cs:23]: This is only called in on place. It’s not a library, just part of an app. It’s less important in that context

File: src/Aspire.Dashboard/Otlp/Http/OtlpHttpEndpointsBuilder.cs [hunk 22]
```diff
@@ -0,0 +1,205 @@
...
+                    }
+                case KnownContentType.Json:
+                default:
+                    context.Response.StatusCode = StatusCodes.Status415UnsupportedMediaType;
+                    return Empty;
+            }
+        }
+    }
+
+    private sealed class OtlpResult<T> : IResult where T : IMessage
```
Comment (captainsafia) [inline src/Aspire.Dashboard/Otlp/Http/OtlpHttpEndpointsBuilder.cs:126]: Is it helpful for this result type to implement `IEndpointMetadataProvider` so that information about the response produced from these endpoints can surface to documentation tools like OpenAPI?

Comment (JamesNK) [reply src/Aspire.Dashboard/Otlp/Http/OtlpHttpEndpointsBuilder.cs:126]: Right now, the dashboard doesn't produce OpenAPI. This can improve later if that changes.

**Comment by JamesNK** [comment]
Merged! Thanks for your work on this @sharpSteff.
